### PR TITLE
CRIMAP-89 add prototype return reasons report

### PR DESCRIPTION
**Pull request by timpeat**
## Description of change
Adds prototype return reasons report.
Refactor temporal reports to use a TimePeriod value object.
Refactor sorting by extracting sortable_struct methods.

## Link to relevant ticket
[CRIMAPP-89](https://dsdmoj.atlassian.net/browse/CRIMAPP-89)

## Notes for reviewer
Requires the following datastore changes and must not be merged until they have been deployed:
https://github.com/ministryofjustice/laa-criminal-applications-datastore/pull/185

### After changes:
Screenshot 2023-11-24 at 10 12 15
Screenshot 2023-11-24 at 10 12 04

## How to manually test the feature

Pull this version of the datastore: https://github.com/ministryofjustice/laa-criminal-applications-datastore/pull/185 .
Log in as a supervisor and click on the "Return reasons report" from the reports dashboard.

[CRIMAPP-89]: https://dsdmoj.atlassian.net/browse/CRIMAPP-89?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/models/concerns/sortable_struct.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    def default_sort_by
+      self::DEFAULT_SORT_BY
+    end
+
+    def default_sort_direction
+      self::DEFAULT_SORT_DIRECTION
+    end
+
+    def new_or_default(params = nil)
```
Comment (hiboabd) [inline app/models/concerns/sortable_struct.rb:22]: Is this the best name?

File: app/models/reporting/caseworker_report_sorting.rb [hunk 2]
```diff
@@ -10,14 +10,8 @@ class CaseworkerReportSorting < Sorting
...
 
-    attribute :sort_direction, Types::SortDirection
-    attribute :sort_by, Types::String.default(DEFAULT_SORT_BY).enum(*SORTABLE_COLUMNS)
-
-    class << self
-      def default
-        new(sort_by: DEFAULT_SORT_BY)
-      end
-    end
+    include SortableStruct
```
Comment (hiboabd) [inline app/models/reporting/caseworker_report_sorting.rb:15]: Nice 🚀

File: app/models/reporting/temporal_report.rb [hunk 3]
```diff
@@ -2,48 +2,50 @@ module Reporting
...
+    attr_reader :time_period, :report_type
 
-    def initialize(date:, report_type:)
-      @date = date || latest_complete_report_date
+    def initialize(time_period:, report_type:)
+      @time_period = time_period
       @report_type = Types::TemporalReportType[report_type]
     end
 
+    delegate :interval, to: :time_period
```
Comment (hiboabd) [inline app/models/reporting/temporal_report.rb:12]: Remove delegate - use `time_period.interval`

File: spec/system/reporting/return_reasons_report_spec.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        expect(cell).to have_content expected_data[i]
+      end
+    end
+  end
+
+  it_behaves_like 'a table with sortable headers' do
+    let(:active_sort_headers) { ['Date returned'] }
+    let(:active_sort_direction) { 'ascending' }
+    let(:inactive_sort_headers) { ['Return reason', 'Reference number', 'Office code'] }
+  end
```
Comment (hiboabd) [inline spec/system/reporting/return_reasons_report_spec.rb:87]: 🚀

File: app/models/concerns/sortable_struct.rb (hunk 1 above)
Comment (timpeat) [reply app/models/concerns/sortable_struct.rb:22]: I think I'd like to refactor the controller where this is used so will take a look then.

**Review by hiboabd — COMMENTED** [review]
👍

**Review by hiboabd — APPROVED** [review]

**Pull request merged**